Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -165,15 +209,356 @@ theorem finprod_heightOneSpectrum_factorization (I : Ideal R) (hI : I ≠ 0) :
+    count K v (I * I') = count K v I + count K v I' := by
+  have hv : Irreducible (Associates.mk v.asIdeal) := by apply v.associates_irreducible
+  obtain ⟨a, J, ha, haJ⟩ := exists_eq_spanSingleton_mul I
+  have ha_ne_zero : Associates.mk (Ideal.span {a} : Ideal R) ≠ 0 := by
+    rw [Ne.def, Associates.mk_eq_zero, Ideal.zero_eq_bot, Ideal.span_singleton_eq_bot]; exact ha
+  have hJ_ne_zero : Associates.mk J ≠ 0 := Associates.mk_ne_zero.mpr (ideal_factor_ne_zero hI haJ)
+  obtain ⟨a', J', ha', haJ'⟩ := exists_eq_spanSingleton_mul I'
+  have ha'_ne_zero : Associates.mk (Ideal.span {a'} : Ideal R) ≠ 0 := by
+    rw [Ne.def, Associates.mk_eq_zero, Ideal.zero_eq_bot, Ideal.span_singleton_eq_bot]; exact ha'
+  have hJ'_ne_zero : Associates.mk J' ≠ 0 :=
+    Associates.mk_ne_zero.mpr (ideal_factor_ne_zero hI' haJ')
+  have h_prod : I * I' = spanSingleton R⁰ ((algebraMap R K) (a * a'))⁻¹ * ↑(J * J') := by
+    rw [haJ, haJ', mul_assoc, mul_comm (J : FractionalIdeal R⁰ K), mul_assoc, ← mul_assoc,
+      spanSingleton_mul_spanSingleton, coeIdeal_mul, RingHom.map_mul, mul_inv,
+      mul_comm (J : FractionalIdeal R⁰ K)]
+  rw [count_well_defined K v hI haJ, count_well_defined K v hI' haJ',
+    count_well_defined K v (mul_ne_zero hI hI') h_prod, ← Associates.mk_mul_mk,
+    Associates.count_mul hJ_ne_zero hJ'_ne_zero hv, ← Ideal.span_singleton_mul_span_singleton,
+    ← Associates.mk_mul_mk, Associates.count_mul ha_ne_zero ha'_ne_zero hv]
+  push_cast
+  ring
+
+/-- For nonzero `I, I'`, `val_v(I*I') = val_v(I) + val_v(I')`. If `I` or `I'` is zero, then
+`val_v(I*I') = 0`. -/
+theorem count_mul' (I I' : FractionalIdeal R⁰ K) :
+    count K v (I * I') = if I ≠ 0 ∧ I' ≠ 0 then count K v I + count K v I' else 0 := by
+  split_ifs with h
+  · exact count_mul K v h.1 h.2
+  · push_neg at h
+    by_cases hI : I = 0
+    · rw [hI, MulZeroClass.zero_mul, count, dif_pos (Eq.refl _)]
+    · rw [h hI, MulZeroClass.mul_zero, count, dif_pos (Eq.refl _)]
+
+/-- val_v(1) = 0. -/
+theorem count_one : count K v (1 : FractionalIdeal R⁰ K) = 0 := by
+  have h1 : (1 : FractionalIdeal R⁰ K) =
+      spanSingleton R⁰ ((algebraMap R K) 1)⁻¹ * ↑(1 : Ideal R) := by
+    rw [(algebraMap R K).map_one, Ideal.one_eq_top, coeIdeal_top, mul_one, inv_one,
+      spanSingleton_one]
+  rw [count_well_defined K v one_ne_zero h1, Ideal.span_singleton_one, Ideal.one_eq_top, sub_self]
+
+theorem count_prod {S : Finset (FractionalIdeal R⁰ K)} (hS : ∀ I ∈ S, I ≠ 0) :
+    count K v (∏ I in S, I) = ∑ I in S, (count K v I) := by
+  induction' S using Finset.induction with i S hi hrec
+  ·  rw [Finset.prod_empty, Finset.sum_empty, count_one]
+  · have hS' : ∀ (I : FractionalIdeal R⁰ K), I ∈ S → I ≠ 0 := fun I hI =>
+      hS I (Finset.mem_insert_of_mem hI)
+    have hS0 : ∏ x in S, x ≠ 0 := Finset.prod_ne_zero_iff.mpr hS'
+    have hi0 : i ≠ 0 := hS i (Finset.mem_insert_self i S)
+    rw [Finset.prod_insert hi, Finset.sum_insert hi, count_mul K v hi0 hS0, hrec hS']
+
```

Comment:
I think something like this would be useful?
```lean
theorem count_coe (J : Ideal R) (hJ : J ≠ 0) :
    count K v J = (Associates.mk v.asIdeal).count (Associates.mk J).factors := by
  classical
  rw [count_well_defined K (J := J) (a := 1), Ideal.span_singleton_one, sub_eq_self,
    Nat.cast_eq_zero, ← Ideal.one_eq_top, Associates.mk_one, Associates.factors_one,
    Associates.count_zero v.associates_irreducible]
  · simpa only [ne_eq, coeIdeal_eq_zero]
  · simp

theorem count_coe_nonneg (J : Ideal R) : 0 ≤ count K v J := by
  by_cases hJ : J = 0
  · simp [hJ, count_zero]
  · simp [count_coe K v J hJ]

theorem count_mono {I J} (hI' : I ≠ 0) (h : I ≤ J) : count K v J ≤ count K v I := by
  by_cases hJ : J = 0
  · exact (hI' (FractionalIdeal.le_zero_iff.mp (h.trans hJ.le))).elim
  have := FractionalIdeal.mul_le_mul_left h J⁻¹
  rw [inv_mul_cancel hJ, FractionalIdeal.le_one_iff_exists_coeIdeal] at this
  obtain ⟨J', hJ'⟩ := this
  rw [← mul_inv_cancel_left₀ hJ I, ← hJ', count_mul K v hJ, le_add_iff_nonneg_right]
  · exact count_coe_nonneg K v J'
  · exact hJ' ▸ mul_ne_zero (inv_ne_zero hJ) hI'
```

Reply:
I added these, thanks!